endaq/calc/shock.py

```diff
@@ -11,71 +12,247 @@
 
-def rel_displ(df: pd.DataFrame, omega: float, damp: float = 0) -> pd.DataFrame:
-    """Calculate the relative displacement for a SDOF system."""
-    # Generate the transfer function
-    #   H(s) = L{z(t)}(s) / L{y"(t)}(s) = (1/s²)(Z(s)/Y(s))
-    # for the PDE
-    #   z" + (2ζω)z' + (ω^2)z = -y"
-    dt = (df.index[-1] - df.index[0]) / (len(df.index) - 1)
-    if isinstance(dt, (np.timedelta64, pd.Timedelta)):
-        dt = dt / np.timedelta64(1, "s")
 
+def _rel_displ_transfer_func(omega: float, damp: float = 0.0, dt: float = 1.0):
+    """
+    Generate the transfer function
+        H(s) = L{z(t)}(s) / L{y"(t)}(s) = (1/s²)(Z(s)/Y(s))
+    for the PDE
+        z" + (2ζω)z' + (ω²)z = -y"
+
+    .. seealso::
+
+        `Pseudo Velocity Shock Spectrum Rules For Analysis Of Mechanical Shock, Howard A. Gaberson <https://info.endaq.com/hubfs/pvsrs_rules.pdf>`_
+
+        `SciPy transfer functions <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.TransferFunction.html>`_
+        Documentation for the transfer function class used to characterize the
+        relative displacement calculation.
+    """
     with warnings.catch_warnings():
         warnings.simplefilter("ignore", scipy.signal.BadCoefficients)
 
-        tf = scipy.signal.TransferFunction(
+        return scipy.signal.TransferFunction(
             [-1],
             [1, 2 * damp * omega, omega ** 2],
         ).to_discrete(dt=dt)
 
-    return df.apply(
+
+def rel_displ(accel: pd.DataFrame, omega: float, damp: float = 0.0) -> pd.DataFrame:
+    """
+    Calculate the relative displacement for a SDOF system.
+
+    The "relative" displacement follows the transfer function:
+        H(s) = L{z(t)}(s) / L{y"(t)}(s) = (1/s²)(Z(s)/Y(s))
+    for the PDE:
+        z" + (2ζω)z' + (ω²)z = -y"
+
+    :param accel: the absolute acceleration y"
+    :param omega: the natural frequency ω of the SDOF system
+    :param damp: the damping coefficient ζ of the SDOF system
+    :return: the relative displacement z of the SDOF system
+
+    .. seealso::
+
+        `Pseudo Velocity Shock Spectrum Rules For Analysis Of Mechanical Shock, Howard A. Gaberson <https://info.endaq.com/hubfs/pvsrs_rules.pdf>`_
+
+        `SciPy transfer functions <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.TransferFunction.html>`_
+        Documentation for the transfer function class used to characterize the
+        relative displacement calculation.
+
+        `SciPy biquad filter <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lfilter.html>`_
+        Documentation for the biquad function used to implement the transfer
+        function.
+    """
+    dt = utils.sample_spacing(accel)
+    tf = _rel_displ_transfer_func(omega, damp, dt)
+
+    return accel.apply(
+        functools.partial(scipy.signal.lfilter, tf.num, tf.den, axis=0),
+        raw=True,
+    )
+
+
+def _abs_accel_transfer_func(omega: float, damp: float = 0.0, dt: float = 1.0):
```

Comment:
Typehint for return

Reply:
see 06f2b0ae993894190bbe6c5aa510cddf75c58bde